### ISSUE TITLE
Where are the negatively paired patches sampled from x_aug?

### ISSUE THREAD

**Issue by DeepTag** [lyndonzheng/F-LSeSim#2]
Thank you for releasing this wonderful code. I am confused with the implementation of the contrastive learning part. In the paper, it mentioned the negatively paired patches whether come from an image y or a different location in the image x_aug. In the code

        if self.opt.augment:
            norm_aug_A, norm_aug_B = self.normalization((self.aug_A + 1) * 0.5), self.normalization((self.aug_B + 1) * 0.5)
            norm_real_A = torch.cat([norm_real_A, norm_real_A], dim=0)
            norm_fake_B = torch.cat([norm_fake_B, norm_aug_A], dim=0)
            norm_real_B = torch.cat([norm_real_B, norm_aug_B], dim=0)
        self.loss_spatial = self.Spatial_Loss(self.netPre, norm_real_A, norm_fake_B, norm_real_B)

        feats_src = net(src, self.attn_layers, encode_only=True)
        feats_tgt = net(tgt, self.attn_layers, encode_only=True)
        if other is not None:
            feats_oth = net(torch.flip(other, [2, 3]), self.attn_layers, encode_only=True)
        else:
            feats_oth = [None for _ in range(n_layers)]

        total_loss = 0.0
        for i, (feat_src, feat_tgt, feat_oth) in enumerate(zip(feats_src, feats_tgt, feats_oth)):
            loss = self.criterionSpatial.loss(feat_src, feat_tgt, feat_oth, i)
            total_loss += loss.mean()

        sam_neg1 = (sim_src.bmm(sim_other.permute(0, 2, 1))).view(-1, Num) / self.T
        sam_neg2 = (sim_tgt.bmm(sim_other.permute(0, 2, 1))).view(-1, Num) / self.T
        sam_self = (sim_src.bmm(sim_tgt.permute(0, 2, 1))).view(-1, Num) / self.T
        sam_self = torch.cat([sam_self, sam_neg1, sam_neg2], dim=-1)
        loss = self.cross_entropy_loss(sam_self, torch.arange(0, sam_self.size(0), dtype=torch.long, device=sim_src.device) % (Num))

It seems like sam_neg1 is the from x and y/y_aug. But sam_neg2 is from y^hat and y or x_aug and y_aug. I didn't see the negatively paired patches from x and x_aug with a different patch center. Is my understanding correct?

**Comment by lyndonzheng** [comment]
Hi @DeepTag, the sam_self between sim_src and sim_tgt is an N*N matrix, where only diagonal items belong to the positive samples, and the other items are  negative samples from x_aug, and \hat{y}. The sam_neg1 and sam_neg2 are used for obtaining more negative samples (N*3N) due to the previous contrastive learning suggested that more negative samples can encourage the converging. Therefore, for every positive paired patches, we sampled another 3*N-1 negative pairs.

**Comment by DeepTag** [comment]
Thanks for clarifying it. Another question is why we have flipping 'other':
        if other is not None:
                feats_oth = net(torch.flip(other, [2, 3]), self.attn_layers, encode_only=True)
            else:
                feats_oth = [None for _ in range(n_layers)]
Is this flipping for avoiding slection of the very similar patches in \hat{y} and y on a same patch center, therefore to ensure these patches are axactly negatively paired?

**Comment by lyndonzheng** [comment]
Hi @DeepTag, yes, in contrastive learning, the negative pairs are automatically selected by the sampling. We can not always ensure those sampled structures are very different. Therefore, we do the 180° rotation to encourage the negative samples are really different with the queried patch.

**Comment by DeepTag** [comment]
That make sense. One further question here. In both the fixed and learned self-similarity methods, we use some layers' features of VGG16 in the code. How to decide which layer to be used?

**Comment by lyndonzheng** [comment]
Hi @DeepTag, please refer to our [website](https://www.chuanxiaz.com/), in which we provide the poster and video for this project, and we show the self-attention map in different layers. We select the multilevel, multiresolution attention map for the structure representation.

**Comment by DeepTag** [comment]
Yes, in your work you used multi-layers' attention maps (you used features from layers relu3_1 and relu4_1). I want to know, how did you determine to select the layers? Why not use other layers' features?
